src/crimson/net/SocketConnection.cc

```diff
@@ -465,7 +471,7 @@ seastar::future<> SocketConnection::handle_connect()
       }
       h.connect_seq = h.connect.connect_seq + 1;
       h.peer_global_seq = h.connect.global_seq;
-      set_features((uint64_t)h.reply.features & (uint64_t)h.connect.features);
+      set_features((uint64_t)policy.features_supported & (uint64_t)h.connect.features);
```

Comment:
i think `h.reply.features` at this moment is 
```c
(h.connect.features & policy.features_supported) |
 policy.features_required
```

and `h.connect.features` should always contain all `policy.features_required`. 

so the existing code should be correct. am i right?

Reply:
From debugging, I can see currently `h.reply.features` is not set yet. Without the fix `memset(&h.reply, 0, sizeof(h.reply))`, it will contain a random value. So a bug happens:
1. Randomly, server doesn't set `CEPH_FEATURE_MSG_AUTH` feature, so it sends message with `ceph_msg_footer_old`.
2. Client tries to read the message, expecting `ceph_msg_footer` (length is different), and failed.
3. The unit test hangs forever ...

Then why I set `h.connect.features & policy.features_supported` here:
I see `line #453` already checked `policy.features_required` with `h.connect.features` before, or tag `CEPH_MSGR_TAG_FEATURES` will be replied instead.
And `line #502` only sets `h.reply.features = policy.features_supported` later.

What do you think?

Reply:
ahh, i am sorry. i thought `send_connect_reply_ready()` got called before we `set_features()`.

you are right!